### ISSUE TITLE
Incorrect Exit Code for Transfer Limit Exceeded

### ISSUE THREAD

**Issue by darthShadow** [rclone/rclone#3203]
#### What is the problem you are having with rclone?

Incorrect Exit Code by rclone when the specified max-transfer limit is reached. As per the docs @ https://rclone.org/docs/#exit-code, the exit-code should be `8` instead of `7` which indicates a fatal error.

This is just my guess from a brief overview of the related code, but I believe that the error stems from https://github.com/ncw/rclone/blob/7d70e9266446901e93a888212edd06765821047b/fs/sync/sync.go#L178-L183 where the actual error is added to which fails the eventual comparison test for the error type.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.47.0-062-g99bc013c-beta
- os/arch: linux/amd64
- go version: go1.12.5
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

`Debian 9 64-bit`

####  Which cloud storage system are you using? (eg Google Drive)

`Google Drive`

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone sync Drive: DriveCopy: --tpslimit=5 --tpslimit-burst=50 --max-transfer=700G`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

Relevant Lines:

```
2019/05/16 07:54:40 ERROR : Cancelling sync due to fatal error: Max transfer limit reached as set by --max-transfer
2019/05/16 07:54:40 INFO  : Google drive root '': Waiting for checks to finish
2019/05/16 07:54:40 INFO  : Google drive root '': Waiting for transfers to finish
2019/05/16 07:54:41 INFO  : o6i2cgd36bmi2nva56ubo4a268/i28c365mh5iu29bbbq5h5bf960/q58f203e7ps6gblq0fu0f36954/o6omqo0u6b7sb4rjhdjf7qsmtg/dajheiohfhb8b7vag018ke
paucfq3fau0kb3l75fqstcii04ch32jcj7p79rnv86jd6v4v8atj3p81pcq460pge65ut8p7n02ptt7eo: Copied (server side copy)
2019/05/16 07:54:41 ERROR : Fatal error received - not attempting retries
2019/05/16 07:54:41 Failed to sync: Max transfer limit reached as set by --max-transfer
```

**Comment by ncw** [comment]
Your analysis looks good!

Probably what should happen is that here

https://github.com/ncw/rclone/blob/cff85f0b9551881be535a4847b56742451e6b476/cmd/cmd.go#L445

The error should be unwrapped then `resolveExitCode` called recursively.

Fancy sending in a pull request?

**Comment by darthShadow** [comment]
Will do shortly. Just a single recursion sounds fine? Also, I guess it should be for the whole switch-case or only for the fatal error case?

**Comment by ncw** [comment]
> Will do shortly. I guess it should be for the whole switch-case or only for the fatal error case?

I was thinking only for the fatal error case, though maybe the Retry/NoRetry cases should be treated the same... Probably what we want to do is attempt to unwrap the fatal error, and if it isn't anything special exit with `exitCodeFatalError` which is a bit more complicated.

**Comment by kadafax** [comment]
Any news on this ? It could be really usefull to check on the exit code n°8 in a script to sleep 24h before attempting a "rclone sync" again.

**Comment by ncw** [comment]
@kadafax would you like to work on a fix? I don't think it should be too complicated (but I could be wrong!)

**Comment by anuar45** [comment]
Hi Nick, can I try?

**Comment by ncw** [comment]
@anuar45 please go ahead 😄

**Comment by anuar45** [comment]
Just checked out first commit, interesting

**Comment by anuar45** [comment]
Nick, can you help me,

Cant understand why [_, unwrapped := fserrors.Cause(err)](https://github.com/rclone/rclone/blob/cff85f0b9551881be535a4847b56742451e6b476/cmd/cmd.go#L429) cant unwrap down to **accounting.ErrorMaxTransferLimitReached** a few lines below.

**Comment by ncw** [comment]
> Nick, can you help me,
> 
> Cant understand why [_, unwrapped := fserrors.Cause(err)](https://github.com/rclone/rclone/blob/cff85f0b9551881be535a4847b56742451e6b476/cmd/cmd.go#L429) cant unwrap down to **accounting.ErrorMaxTransferLimitReached** and match it few lines below.

Not sure...

I'd put a `fmt.Printf("err = %#v\n", err)` in there to find out what kind of error it really is.

**Comment by anuar45** [comment]
Thanks Nick,

This what I found
unwrapped err = Max transfer limit reached as set by --max-transfer, type = *errors.fundamental

and of course it doesn't match to accounting.ErrorMaxTransferLimitReached.

Maybe switch to err.Unwrap(), instead of err.Cause() method and use new errors.Is(err, accounting.ErrorMaxTransferLimitReached) func from stdlib.

Or make custom errors.Is(err, targeterr error) in lib/errors/errors.go.

**Comment by anuar45** [comment]
Nick please, check also my last comment.

Now I see we should comply with go 1.11 and 1.12, so in this case i think we should implement our own errors.Is().

**Comment by ncw** [comment]
I see...

The definition is

https://github.com/rclone/rclone/blob/ce23cb209313d870141b1c062f7c2a3bc8cf7f3e/fs/accounting/accounting.go#L21

Perhaps we should split that up into

```go
var ErrorMaxTransferLimitReached = errors.New("Max transfer limit reached as set by --max-transfer")
var ErrorMaxTransferLimitReachedFatal = fserrors.FatalError(ErrorMaxTransferLimitReached)
```

And check for `ErrorMaxTransferLimitReached` but return `ErrorMaxTransferLimitReachedFatal`

**Comment by anuar45** [comment]
Thanks you for your comments, your solution looks more natural

New commit ready.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52

Thank you :-)
